prov/gni/src/gnix_vc.c

```diff
@@ -0,0 +1,1156 @@
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/*
+ * code for managing VC's
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+
+#include "gnix.h"
+#include "gnix_vc.h"
+#include "gnix_util.h"
+#include "gnix_datagram.h"
+#include "gnix_cm_nic.h"
+#include "gnix_mbox_allocator.h"
+#include "gnix_hashtable.h"
+
+/*******************************************************************************
+ * Helper functions.
+ ******************************************************************************/
+
+static int __gnix_vc_free_id(struct gnix_vc *vc)
+{
+	struct gnix_fid_ep *ep = NULL;
+	struct gnix_nic *nic;
+
+	ep = vc->ep;
+	if (ep == NULL)
+		return -FI_EINVAL;
+
+	nic = ep->nic;
+	if (nic == NULL)
+		return -FI_EINVAL;
+
+	__gnix_clear_bit(&nic->vi_id_bitmap, vc->vc_id);
+
+	return FI_SUCCESS;
+}
+
+/*
+ * this function is needed to allow for quick lookup of a vc based on
+ * the contents of the GNI CQE coming off of the GNI RX CQ associated
+ * with GNI nic being used by this VC.  Using a bitmap to expedite
+ * scanning vc's in the case of a GNI CQ overrun.
+ */
+
+static int _gnix_vc_get_id(struct gnix_vc *vc)
+{
+	int ret = FI_SUCCESS;
+	struct gnix_fid_ep *ep = NULL;
+	struct gnix_nic *nic;
+	struct gnix_vc **table_base;
+
+	ep = vc->ep;
+	if (ep == NULL)
+		return -FI_EINVAL;
+
+	nic = ep->nic;
+	if (nic == NULL)
+		return -FI_EINVAL;
+
+	/*
+	 * TODO:  really need to search bitmap for clear
+	 * bit before resizing the table
+	 */
+
+	fastlock_acquire(&nic->vc_id_lock);
+	if (nic->vc_id_table_capacity == nic->vc_id_table_count) {
+		table_base = realloc(nic->vc_id_table,
+				     2 * nic->vc_id_table_capacity *
+				     sizeof(struct gnix_vc *));
+		if (table_base == NULL) {
+			ret =  -FI_ENOMEM;
+			goto err;
+		}
+		nic->vc_id_table_capacity *= 2;
+		nic->vc_id_table = table_base;
+	}
+
+	nic->vc_id_table[nic->vc_id_table_count] = vc;
+	vc->vc_id = nic->vc_id_table_count;
+
+	/*
+	 * set bit in the bitmap
+	 */
+
+	__gnix_set_bit(&nic->vi_id_bitmap, nic->vc_id_table_count);
```

Comment:
ah, __gnix_set_bit is what's in the gnix_bitmap.h.  Should we change these to be _gnix_?  It would seem that since they are actually in another file - even though its a header file - that they ought to be single '_'.